### PR TITLE
C++: Navio: Common: gpio: detect RPi version & gpio_base

### PR DESCRIPTION
**Pull request by corteccia**
Hello again! 

Recently I faced with problem with non-working LED example in C++ on Raspberry Pi 4. I found out that RPi 4 has another GPIO_BASE and was not defined in existed gpio.cpp (there were only three versions). It couldn't be simply defined because RPi version was detected using Hardware info but RPi 3 and 4 have the same hardware (BCM2835).

This is why I updated the mechanism of RPi version detection and gpio bases. 

After this change, you can be able to run C++ LED example on RPi 4 and it would work.

### CODE REVIEW

File: C++/Navio/Common/gpio.cpp [hunk 1]
```diff
@@ -16,9 +16,8 @@
 #define HIGH                1
 
 /* Raspberry Pi GPIO memory */
-#define BCM2708_PERI_BASE   0x20000000
-#define BCM2709_PERI_BASE   0x3F000000
-#define BCM2835_PERI_BASE   0x3F000000
+#define BCM2708_PERI_BASE   0x3F000000
+#define BCM2711_PERI_BASE   0xfe000000
```
Comment (AlexeyBulatov) [inline C++/Navio/Common/gpio.cpp]:
```suggestion
#define BCM2711_PERI_BASE   0xFE000000
```
Don't mix

File: C++/Navio/Common/gpio.cpp [hunk 2]
```diff
@@ -137,37 +129,36 @@ void Pin::toggle()
...
-            } else if (strstr(buffer, v2) != NULL) {
-                fclose(fd);
-                return 2;
-            } else if (strstr(buffer, v1) != NULL) {
-                fclose(fd);
-                return 1;
-            }
+    int rpi_version;
+
+    FILE *f = fopen("/sys/firmware/devicetree/base/model", "r");
```
Comment (AlexeyBulatov) [inline C++/Navio/Common/gpio.cpp]: Let's add define for this path.

File: C++/Navio/Common/gpio.cpp [hunk 3]
```diff
@@ -137,37 +129,36 @@ void Pin::toggle()
...
+    if (f != 0 && fgets(buffer, MAX_SIZE_LINE, f) != 0) {
+        int ret = sscanf(buffer + 12, "%d", &rpi_version);
+        fclose(f);
```
Comment (AlexeyBulatov) [inline C++/Navio/Common/gpio.cpp]: You don't close file if the condition is not met.

File: C++/Navio/Common/gpio.cpp [hunk 4]
```diff
@@ -137,37 +129,36 @@ void Pin::toggle()
...
+        if (ret == -1) {
```
Comment (AlexeyBulatov) [inline C++/Navio/Common/gpio.cpp]: EOF

File: C++/Navio/Common/gpio.cpp [hunk 5]
```diff
@@ -137,37 +129,36 @@ void Pin::toggle()
...
-                return 2;
-            } else if (strstr(buffer, v1) != NULL) {
-                fclose(fd);
-                return 1;
-            }
+    int rpi_version;
+
+    FILE *f = fopen("/sys/firmware/devicetree/base/model", "r");
+    if (f != 0 && fgets(buffer, MAX_SIZE_LINE, f) != 0) {
+        int ret = sscanf(buffer + 12, "%d", &rpi_version);
```
Comment (AlexeyBulatov) [inline C++/Navio/Common/gpio.cpp]: Magic number. Are we sure that this offset will always be the same?

**Review by AlexeyBulatov — CHANGES_REQUESTED** [review]
The commit's message does not match the content. You did more.
And add description, please.

File: C++/Navio/Common/gpio.cpp (hunk 5 above)
Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: It should be because it's common for all RPi's versions, 
`cat /sys/firmware/devicetree/base/model` gives you something like this:
`Raspberry Pi 4 Model B Plus Rev 1.3`

It is changed only version, model and revision, so the offset would always be there.

Comment (AlexeyBulatov) [reply C++/Navio/Common/gpio.cpp]: Magic number. Add define then

**Comment by corteccia** [comment]
I've done all the suggested changes. Is everything good now?

File: C++/Navio/Common/gpio.cpp [hunk 6]
```diff
@@ -32,12 +31,14 @@
 #define GPIO_GET(g)         (*(_gpio+13)&(1<<g)) // 0 if LOW, (1<<g) if HIGH
 
 #define MAX_SIZE_LINE       50
+#define RPI_SIZE            12
```
Comment (AlexeyBulatov) [inline C++/Navio/Common/gpio.cpp]: Bad name. I would not understand what this means without knowing the context.

Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: Is now better?

File: C++/Navio/Common/gpio.cpp [hunk 7]
```diff
@@ -66,15 +67,8 @@ bool Pin::init()
...
-    uint32_t address;
-    int version = getRaspberryPiVersion();
-    if (version == 1) {
-        address = GPIO_BASE(BCM2708_PERI_BASE);
-    } else if (version == 2) {
-        address = GPIO_BASE(BCM2709_PERI_BASE);
-    } else if (version == 3) {
-        address = GPIO_BASE(BCM2835_PERI_BASE);
-    }
+    int address;
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: Use unsigned types for addresses. There's also size_t, uintptr_t etc.

File: C++/Navio/Common/gpio.cpp [hunk 8]
```diff
@@ -92,9 +86,9 @@ bool Pin::init()
 
     /* No need to keep mem_fd open after mmap */
     if (close(mem_fd) < 0) {
-        warn("cannot close mem_fd");   
+        warn("cannot close mem_fd");
         return false;
-    } 
+    }
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp:119]: let's move changes with formatting to another commit.

File: C++/Navio/Common/gpio.cpp [hunk 9]
```diff
@@ -32,12 +31,14 @@
 #define GPIO_GET(g)         (*(_gpio+13)&(1<<g)) // 0 if LOW, (1<<g) if HIGH
 
 #define MAX_SIZE_LINE       50
+#define RPI_NAME_LENGTH     12
 
+#define MODELPATH           "/sys/firmware/devicetree/base/model"
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: `SYSFS_RPI_MODEL_FILE_PATH` ? Defines should be explicit.

File: C++/Navio/Common/gpio.cpp [hunk 10]
```diff
@@ -16,9 +16,8 @@
 #define HIGH                1
 
 /* Raspberry Pi GPIO memory */
-#define BCM2708_PERI_BASE   0x20000000
-#define BCM2709_PERI_BASE   0x3F000000
-#define BCM2835_PERI_BASE   0x3F000000
+#define BCM2708_PERI_BASE   0x3F000000
+#define BCM2711_PERI_BASE   0xFE000000
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: This is not obvious to me why we removed one define and changed current values. We can move it to separate commit and add proper description.

File: C++/Navio/Common/gpio.cpp [hunk 11]
```diff
@@ -137,37 +131,37 @@ void Pin::toggle()
...
         }
     }
 
-    /* defaults to 1 */
-    fprintf(stderr, "Could not detect RPi version, defaulting to 1\n");
-    fclose(fd);
-    return 1;
+    fclose(f);
+
+    return rpi_version;
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: We use C++, not C. Things such as `RPI_NAME_LENGTH` are _**very**_ bad.
I've not tested it a lot and it's ugly, but you can get the idea from the code below. Just create a file 'test.cpp', paste and compile.
```
#include <iostream> // 1 2 3
#include <fstream>
#include <string>

int main(){
    std::size_t i = 0, j = 0;
    std::string line;
    std::ifstream f("test.cpp");
    if (f.is_open() && std::getline(f,line)){
	    for ( ; i < line.length(); i++ ){if (isdigit(line[i])) break;}
	    char ch = line[i];
	    j = atoi(&ch);
	    std::cout << j << std::endl;
	    f.close();

	    // compare the version here
	    return 0;
    }
    std::cout << "unable to get the contents" << std::endl;
    return 1;
}

**Comment by corteccia** [comment]
@AlexanderDranitsa Thank you for the review! Please check it now.

File: C++/Navio/Common/gpio.cpp [hunk 12]
```diff
@@ -137,37 +132,42 @@ void Pin::toggle()
...
-        }
+    size_t rpi_version = 0;
+
+    std::size_t i = 0;
+    std::string buffer;
+    std::ifstream f(SYSFS_RPI_MODEL_FILE_PATH);
+
+    if (f.is_open() && std::getline(f,buffer)){
+      for ( ; i < buffer.length(); i++ ){if (isdigit(buffer[i])) break;}
+      char ret = buffer[i];
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: ret? bad name

File: C++/Navio/Common/gpio.cpp [hunk 13]
```diff
@@ -137,37 +132,42 @@ void Pin::toggle()
...
+
+    switch (rpi_version) {
+      case 1:
+        return BCM2708_PERI_BASE;
+      case 2:
+        return BCM2709_PERI_BASE;
+      case 3:
+        return BCM2835_PERI_BASE;
+      default:
+        warnx("Unexpected version. Set PERI_BASE to default as for RPI2/3\n");
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: `Set PERI_BASE to default as for RPI2/3` I was confused about this, please write such that it's obvious it's you set PERI_BASE to default. Not like you are asking someone to do this. `PERI_BASE` should not be in a warning.

File: C++/Navio/Common/gpio.h [hunk 14]
```diff
@@ -47,7 +47,8 @@ class Pin {
     void    toggle();
 
 private:
-    int getRaspberryPiVersion() const;
+    int getPeriBase() const;
+    size_t getRaspberryPiVersion() const;
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.h]: no reason to use size_t here

**Comment by corteccia** [comment]
Check it now, please.

File: C++/Navio/Common/gpio.cpp [hunk 15]
```diff
@@ -137,37 +132,48 @@ void Pin::toggle()
...
+        return BCM2708_PERI_BASE;
+      case 2:
+        return BCM2709_PERI_BASE;
+      case 3:
+        return BCM2835_PERI_BASE;
+      default:
+        fprintf(stderr, "Unexpected version. Setting up configurations as on RPI2/3\n");
+        return BCM2709_PERI_BASE;
+}
+}
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: codestyle

File: C++/Navio/Common/gpio.cpp [hunk 16]
```diff
@@ -137,37 +132,48 @@ void Pin::toggle()
...
-                fclose(fd);
-                return 1;
-            }
+    int rpi_version;
+
+    size_t i = 0;
+    std::string buffer;
+    std::ifstream f(SYSFS_RPI_MODEL_FILE_PATH);
+
+    if (f.is_open()){
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: codestyle

File: C++/Navio/Common/gpio.cpp [hunk 17]
```diff
@@ -137,37 +132,48 @@ void Pin::toggle()
...
+        rpi_version = atoi(&version_char);
+
+        return rpi_version;
+     }
+        fprintf(stderr, "Could not read RPi model, defaulting to 1\n");
+        return 1;
+     }
+
+     fprintf(stderr, "Could not open ", SYSFS_RPI_MODEL_FILE_PATH, "defaulting to 1\n");
+     return 1;
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: 1. Invert the logic (random [link](https://www.geeksforgeeks.org/writing-clean-else-statements/) from Google).
2. "defaulting to" - bad decision. Just drop it.

Comment (AlexanderDranitsa) [reply C++/Navio/Common/gpio.cpp]: 3. return 1 - bad decision. Add a `#define` for exit status.

Comment (AlexanderDranitsa) [reply C++/Navio/Common/gpio.cpp]: 4. Why fprintf? Am I wrong or `std::cerr <<` is how we do it in CPP?

File: C++/Navio/Common/gpio.cpp [hunk 18]
```diff
@@ -137,37 +132,48 @@ void Pin::toggle()
...
+    int rpi_version;
+
+    size_t i = 0;
+    std::string buffer;
+    std::ifstream f(SYSFS_RPI_MODEL_FILE_PATH);
+
+    if (f.is_open()){
+      while (std::getline(f, buffer)) {
+        while (!isdigit(buffer[i])) { //first digit is a number of RPI version (from 1 to 4)
+          i++; //skip
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: This is bad. There can be _no_ digits in the common case. we do not know this. Adding a potentially infinite loop is a bad decision.

File: C++/Navio/Common/gpio.cpp [hunk 19]
```diff
@@ -137,20 +136,19 @@ uint32_t Pin::getPeriBase() const
     int rpi_version = getRaspberryPiVersion();
 
     switch (rpi_version) {
-      case 1:
+    case 1:
+    case 2:
+    case 3:
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: case x ... y:

File: C++/Navio/Common/gpio.cpp (hunk 17 above)
Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: 2. Shouldn't I inform users which model I will use if there would be an error?
3. Is `DEFAULT_RPI_VERSION` good for this define?
4. Yes, it just was in original code and I followed it.

File: C++/Navio/Common/gpio.cpp (hunk 18 above)
Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: Can I not use `break` and just doing it like this:
```cpp
while (i < buffer.length()) {
          if (isdigit(buffer[i])) {
            char version_char = buffer[i];
            rpi_version = atoi(&version_char);

            return rpi_version;
          }
          i++;
        }
```

File: C++/Navio/Common/gpio.h [hunk 20]
```diff
@@ -47,6 +47,7 @@ class Pin {
     void    toggle();
 
 private:
+    int getPeriBase() const;
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.h]: What's with the return type? Did you compile it before push?

Comment (corteccia) [reply C++/Navio/Common/gpio.h]: Thank you! I miss that while committing.

File: C++/Navio/Common/gpio.cpp (hunk 17 above)
Comment (AlexanderDranitsa) [reply C++/Navio/Common/gpio.cpp]: If there would be an error you do nothing but tell the user about it and exit with an error.

File: C++/Navio/Common/gpio.cpp (hunk 18 above)
Comment (AlexanderDranitsa) [reply C++/Navio/Common/gpio.cpp]: So, it looks like the `for` loop, isn't it?

Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: For sure, and without `break`

File: C++/Navio/Common/gpio.cpp (hunk 17 above)
Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: Get it. I thought we don't exit with it, just setting something else.

Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: 1. Look for it, please. I change the logic in the first case (f.is_open) but left the second one as it is because there's one common error. Should I invert it too?

**Comment by AlexanderDranitsa** [comment]
@corteccia 
Just looked through the code once again and thought we probably can do smth like this, I like it more
```
if (!f){
    std::cout << "Unable get contents" << std::endl;
    return 1;
}
    
std::string s;
int model_version = 0;
while(f >> s) {
    try {
        model_version = stoi(s);
        break;
    }
    catch(std::invalid_argument) {}
}
```

**Comment by corteccia** [comment]
@AlexanderDranitsa Okay, it seems really better. Updated that, check, please.

File: C++/Navio/Common/gpio.cpp [hunk 21]
```diff
@@ -137,37 +134,45 @@ void Pin::toggle()
...
+    int rpi_version = getRaspberryPiVersion();
+
+    switch (rpi_version) {
+    case 1 ... 3:
+        return BCM2708_PERI_BASE;
+    case 4:
+        return BCM2711_PERI_BASE;
+    default:
+        std::cerr << "Unexpected version. Setting up configurations as on RPI2/3" << std::endl;
+        return BCM2708_PERI_BASE;
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: Unexepted version -> error, exit.

**Comment by AlexanderDranitsa** [comment]
Can't find any pushes to check.

**Comment by AlexanderDranitsa** [comment]
> @AlexanderDranitsa Okay, it seems really better. Updated that, check, please.

Can't find any pushes to check.

**Comment by corteccia** [comment]
@AlexanderDranitsa this is probably github problem again: you can see it in my profile for `led-fix` branch but it didn't update it here for some reason, we need to wait.

File: C++/Navio/Common/gpio.cpp [hunk 22]
```diff
@@ -137,37 +135,47 @@ void Pin::toggle()
...
-    return 1;
+    while (f >> buffer) {
+      try {
+        rpi_version = stoi(buffer);
+        return rpi_version;
+      }
+      catch(std::invalid_argument) {}
+    }
+
+    f.close();
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: Oh, and we can omit this. RAII deals with it better.

File: C++/Navio/Common/gpio.cpp [hunk 23]
```diff
@@ -137,37 +135,47 @@ void Pin::toggle()
...
     }
 
-    /* defaults to 1 */
-    fprintf(stderr, "Could not detect RPi version, defaulting to 1\n");
-    fclose(fd);
-    return 1;
+    while (f >> buffer) {
+      try {
+        rpi_version = stoi(buffer);
+        return rpi_version;
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: Looks better! Now let's just return stoi(buffer)?

File: C++/Navio/Common/gpio.cpp (hunk 22 above)
Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: Sure, thank you!

Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: I pushed these last changes but there's still the same Github problem: it doesn't display them here.

File: C++/Navio/Common/gpio.cpp [hunk 24]
```diff
@@ -9,12 +9,17 @@
...
+#include <fstream>
+#include <string>
+#include <iostream>
 
 #include "gpio.h"
 
 #define LOW                 0
 #define HIGH                1
 
+#define ERROR               1
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: Way too common naming, remember, it's a preprocessor, we can already have `ERROR` define somewhere. Make more concrete naming to avoid confuses.

Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: Would `EXIT_ERROR` be better?

Comment (AlexanderDranitsa) [reply C++/Navio/Common/gpio.cpp]: Yes, better, but if you bind it with our product name and/or process name (we're getting RPi revision) it will be definitely safe and won't be used by mistake.

Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: Good! I changed it to `RPI_MODEL_ERROR`, I think it seems logical.

File: C++/Navio/Common/gpio.cpp [hunk 25]
```diff
@@ -137,37 +135,42 @@ void Pin::toggle()
...
+        return BCM2708_PERI_BASE;
+      case 2:
+        return BCM2709_PERI_BASE;
+      case 3:
+        return BCM2835_PERI_BASE;
+      default:
+        std::cerr << "Unexpected version. Setting up configurations as on RPI2/3" << std::endl;
+        return BCM2709_PERI_BASE;
+    }
+}
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: You add this function at this commit and change your current changes in the next one. Won't it be better to swap commits? Fix addresses before this?

File: C++/Navio/Common/gpio.cpp [hunk 26]
```diff
@@ -67,14 +72,7 @@ bool Pin::init()
...
     uint32_t address;
-    int version = getRaspberryPiVersion();
-    if (version == 1) {
-        address = GPIO_BASE(BCM2708_PERI_BASE);
-    } else if (version == 2) {
-        address = GPIO_BASE(BCM2709_PERI_BASE);
-    } else if (version == 3) {
-        address = GPIO_BASE(BCM2835_PERI_BASE);
-    }
+    address = GPIO_BASE(getPeriBase());
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: So you can get `RPI_MODEL_ERROR` here. Which is `1`. Fix it, please.

File: C++/Navio/Common/gpio.cpp [hunk 27]
```diff
@@ -9,12 +9,17 @@
...
+#include <fstream>
+#include <string>
+#include <iostream>
 
 #include "gpio.h"
 
 #define LOW                 0
 #define HIGH                1
 
+#define RPI_MODEL_ERROR     1
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: Well, the name is good now, thanks! But the `1` is bad. You actually have such a variant of RPi model. And your switch has such `case`. Think about how to fix it, please. As an easy decision - you can read [this](https://en.wikipedia.org/wiki/Hexspeak). We sometimes use it. But it will be good if you find a better solution.

Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: What if I would use `exit(EXIT_FAILURE);` instead of this define? I think it is much better to just exit the code if there's error.

Comment (AlexanderDranitsa) [reply C++/Navio/Common/gpio.cpp]: You can try, but:
- You should exit at only once place
- It's better to exit at a higher level

Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: What do you mean by "exit at a higher level"? 

I was testing it with a test file instead of model_path, using different possibilities and It seems to work fine. But if you say this is unsafe, I will try to find another method.

Comment (AlexanderDranitsa) [reply C++/Navio/Common/gpio.cpp]: Like if you have 3 functions a(), b(), c(), you call `b` and `c` from `a`, it's better to exit at `a` if smth goes wrong at `b` or `c`. If you will add exit directly to `b` and `c`, it will be harder to track where your program can exit. Harder to read and debug. And imagine one day you'll have to add 10 more functions with exits.
If that point doesn't seem critical to you, I have one more and it's a technical one - when you call exit & Co, you will not go through stack unwilling. It's a C function and it doesn't know anything about how CPP and RAII work. You'll get no destructors called. In our example, it means ifstream's state left is undefined (ofc Linux will take care of it, but nevertheless).
Safe options: `return`, `throw & catch`. The second one looks like overkill here and I don't think the `main` has an exception handling here, so I would suggest using returns. Just check them and find a way to tell other parts of the program that smth failed.

Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: Thank you for this explanation! This is my solution: I define `RPI_MODEL_ERROR` as 0 and just would check it in _init_ function. I guess this pretty easy way would be enough. Check it, please.

File: C++/Navio/Common/gpio.cpp [hunk 28]
```diff
@@ -67,15 +72,13 @@ bool Pin::init()
...
-    } else if (version == 2) {
-        address = GPIO_BASE(BCM2709_PERI_BASE);
-    } else if (version == 3) {
-        address = GPIO_BASE(BCM2835_PERI_BASE);
+
+    if (getPeriBase() == 0) {
+      return false;
     }
 
+    address = GPIO_BASE(getPeriBase());
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: Why calling `getPeriBase` twice? it will end up calling `getRaspberryPiVersion`twice, reading through that file twice etc.

Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: Sure! Fixed that.

Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: I fixed typos, therefore I don't know how to better handle `cerr` in switch because it doesn't seem good to output two different errors or zero errors if it would find the same number as `RPI_MODEL_ERROR` in case not using a magic number you've suggested (in our case it can even be some decimal number but it doesn't look good).

File: C++/Navio/Common/gpio.cpp [hunk 29]
```diff
@@ -137,6 +136,22 @@ void Pin::toggle()
...
+    switch (rpi_version) {
+      case 1 ... 3:
+        return BCM2708_PERI_BASE;
+      case 4:
+        return BCM2711_PERI_BASE;
+      default:
+        std::cerr << "Unexpected version." << std::endl;
+        return false;
+    }
+}
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: Move it to the next commit. This commit is just about to fix addresses.
And please check the ret types again.

Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: Moved it, there shouldn't be ret problems now.

**Comment by corteccia** [comment]
@AlexanderDranitsa  I changed it the way we've discussed. I hope it is the final version!

File: C++/Navio/Common/gpio.cpp [hunk 30]
```diff
@@ -67,14 +71,32 @@ bool Pin::init()
...
+          break;
+        case 2:
+        case 3:
+          address = GPIO_BASE(BCM2709_PERI_BASE);
+          break;
+        case 4:
+          address = GPIO_BASE(BCM2711_PERI_BASE);
+          break;
+        default:
+          std::cerr << "Unexpected version" << std::endl;
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]:
```suggestion
          std::cerr << "Unexpected version: " << rpi_version << std::endl;
```

Comment (corteccia) [reply C++/Navio/Common/gpio.cpp]: Sure! Made it

**Comment by AlexanderDranitsa** [comment]
@AlexeyBulatov 
what do you think about it?

**Review by AlexeyBulatov — APPROVED** [review]

File: C++/Navio/Common/gpio.cpp [hunk 31]
```diff
@@ -67,14 +71,32 @@ bool Pin::init()
...
+            break;
+        case 2:
+        case 3:
+            address = GPIO_BASE(BCM2837_PERI_BASE);
+            break;
+        case 4:
+            address = GPIO_BASE(BCM2711_PERI_BASE);
+            break;
+        default:
+            std::cerr << "Unexpected version:" << rpi_version << std::endl;
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]:
```suggestion
            std::cerr << "Unexpected version: " << rpi_version << std::endl;
```

File: C++/Navio/Common/gpio.cpp [hunk 32]
```diff
@@ -16,9 +16,10 @@
 #define HIGH                1
 
 /* Raspberry Pi GPIO memory */
-#define BCM2708_PERI_BASE   0x20000000
-#define BCM2709_PERI_BASE   0x3F000000
-#define BCM2835_PERI_BASE   0x3F000000
+
```
Comment (AlexanderDranitsa) [inline C++/Navio/Common/gpio.cpp]: Why is the blank line added? Let's remove it. Other Comments are not separated via blank lines.

**Review by AlexanderDranitsa — APPROVED** [review]
